src/libraries/System.Runtime/tests/System.IO.FileSystem.Tests/File/OpenHandle.cs

```diff
@@ -95,5 +95,36 @@ public void DeleteOnClose_FileDeletedAfterSafeHandleDispose(FileOptions options)
             }
             Assert.False(File.Exists(path));
         }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void PreallocationSizeVeryLargeThrowsCorrectHResult()
+        {
+            const long VeryLargeFileSize = (long)128 * 1024 * 1024 * 1024 * 1024; // 128TB, large but still allowed by NTFS
+            const int ERROR_DISK_FULL = unchecked((int)0x80070070);
+
+            string path = GetTestFilePath();
+            if (!IOServices.IsDriveNTFS(Path.GetPathRoot(path)))
+            {
+                // Skip the test for non-NTFS filesystems
+                return;
+            }
+
+            if (new DriveInfo(path).TotalFreeSpace >= VeryLargeFileSize)
+            {
+                // Skip the test if somehow the drive is really big.
+                return;
+            }
+
+            try
+            {
+                using (File.OpenHandle(path, mode: FileMode.Create, access: FileAccess.ReadWrite, preallocationSize: VeryLargeFileSize)) { }
+                Assert.Fail("The test should fail due to failure to preallocate a very large file.");
```

Comment:
I can see that this test fails in the CI. I was able to reproduce it locally and debug it.

The sys-call for setting the pre-allocation size fails here:

https://github.com/dotnet/runtime/blob/06a6ea78fbe99fd86f2f46df9b046cb4c4ea18a7/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs#L130-L142

but the error code I keep getting is 87 (ERROR_INVALID_PARAMETER). We need to dive deeper and find out why it's happening.

Reply:
I can see that before #59338 we were using a different sys-call that was most likely using given error codes (ERROR_DISK_FULL and ERROR_FILE_TOO_LARGE), after switching to `SetFileInformationByHandle` we no longer get those errors.

So we need to fix the error handling logic.